allennlp/semparse/worlds/world.py

```diff
@@ -245,71 +276,86 @@ def get_logical_form(self,
+
+        For example, we could get a node with type ``c``, and an action sequence that begins with
+        ``c -> [<r,c>, r]``.  This method will add two children to the input node, consuming
+        actions from the action sequence for nodes of type ``<r,c>`` (and all of its children,
+        recursively) and ``r`` (and all of its children, recursively).  This method assumes that
+        action sequences are produced `depth-first`, so all actions for the subtree under ``<r,c>``
+        appear before actions for the subtree under ``r``.  If there are any actions in the action
+        sequence after the ``<r,c>`` and ``r`` subtrees have terminated in leaf nodes, they will be
+        returned.
+        """
+        if not remaining_actions:
+            logger.error("No actions left to construct current node: %s", current_node)
+            raise ParsingError("Incomplete action sequence")
+        left_side, right_side = remaining_actions.pop(0)
+        if left_side != current_node.label:
+            logger.error("Current node: %s", current_node)
+            logger.error("Next action: %s -> %s", left_side, right_side)
+            logger.error("Remaining actions were: %s", remaining_actions)
+            raise ParsingError("Current node does not match next action")
+        if right_side[0] == '[':
+            # This is a non-terminal expansion, with more than one child node.
+            for child_type in right_side[1:-1].split(', '):
+                if child_type.startswith("'lambda"):
+                    # We need to special-case the handling of lambda here, because it's handled a
+                    # bit weirdly in the action sequence.  This is stripping off the single quotes
+                    # around something like `'lambda x'`.
+                    child_type = child_type[1:-1]
+                child_node = _TreeNode(child_type)
+                current_node.children.append(child_node)
+                if not self.is_terminal(child_type):
+                    remaining_actions = self._construct_node_from_actions(child_node,
+                                                                          remaining_actions,
+                                                                          add_var_function)
+        elif self.is_terminal(right_side):
+            # The current node is a pre-terminal; we'll add a single terminal child.  We need to
+            # check first for whether we need to add a (var _) around the terminal node, though.
+            if add_var_function and right_side in self._lambda_variables:
+                right_side = f"(var {right_side})"
+            if add_var_function and right_side == 'var':
+                raise ParsingError('add_var_function was true, but action sequence already had var')
+            current_node.children.append(_TreeNode(right_side))
+        else:
+            # We've got a unary non-terminal expansion here.  Not very common, but possible.
```

Comment:
Can you give an example of such an expansion?

Reply:
No; we don't have any currently, so this code path isn't actually touched in the tests.  But it's theoretically possible, so I left this in.

Reply:
So that's like `NT1 -> NT2` and `NT2 -> T` or `NT2 -> [NT3, NT4]`? Why would that ever be the case? We could just have `NT1 -> T` or `NT1 -> [NT3, NT4]` instead.

Reply:
Do you want to disallow unary productions?  We can.  I'm just not sure that we should.

Reply:
I cannot think of any cases where we need them. I think we should do that in the interest of making the action space smaller. May not make a huge difference, but still..

Reply:
To be clear: when you're designing your action space, you can choose not to include any unary actions (and neither the wikitables nor the nlvr type declarations have them).  What disallowing them here means is that no one can choose to include unary actions, because we disallow them in the base class.  Are you sure that's what we should do?

Reply:
Right. I meant that I cannot think of a reason anyone would want to include unary actions in their action space, since we can always convert it into an equally expressive action space that does not contain the non-terminals that appear on the right side of unary non-terminal productions in the original action space. So, I think it is alright to disallow users from defining unary non-terminal productions. Do you see any problems with that?